File: programmers/lv2_review/lv_2.12.py
# 소수 찾기
def solution(numbers):
    def solve(s, n, k):
        if k == n:
            res.add(int(s))
            return

        for i in range(len(numbers)):
            if not visited[i]:
                visited[i] = True
                solve(s + numbers[i], n, k + 1)
                visited[i] = False

    cnt = 0

    visited = [False] * len(numbers)
    res = set()
    for i in range(1, len(numbers) + 1):
        solve("", i, 0)
    primes = get_prime_list(10000000)
    for i in res:
        if i in primes:
            cnt += 1

    return cnt


def get_prime_list(n):
    sieve = [True] * n
    for i in range(2, int(n ** 0.5) + 1):
        if sieve[i]:
            for j in range(i + i, n, i):
                sieve[j] = False
    return set([i for i in range(2, n) if sieve[i]])


print(get_prime_list(10000000))

File: programmers/lv2_review/lv2_12.js
// programmers 거리두기 확인하기
const dy = [-1, 1, 0, 0];
const dx = [0, 0, -1, 1];
const INF = 987654321;

function solution(places) {
  const answer = [];
  for (let i = 0; i < places.length; i++) {
    answer.push(check(places[i]) ? 1 : 0);
  }
  return answer;
}

function check(place) {
  let flag = true;
  for (let i = 0; i < 5; i++) {
    for (let j = 0; j < 5; j++) {
      if (place[i][j] === "P") {
        if (!bfs(i, j, place)) {
          flag = false;
          break;
        }
      }
    }
    if (!flag) break;
  }
  return flag;
}

function bfs(sy, sx, place) {
  const q = [];
  const visited = [];
  for (let i = 0; i < 5; i++) {
    const temp = [];
    for (let j = 0; j < 5; j++) {
      temp.push(false);
    }
    visited.push(temp);
  }
  q.push([sy, sx, 0]);
  visited[sy][sx] = true;
  while (q.length) {
    const [y, x, cnt] = q.shift();
    if (sy !== y && sx !== x && place[y][x] === "P") return false;
    if (cnt >= 2) continue;
    for (let i = 0; i < 4; i++) {
      const ny = y + dy[i];
      const nx = x + dx[i];
      if (ny < 0 || ny >= 5 || nx < 0 || nx >= 5) continue;
      if (place[ny][nx] === "X") continue;
      if (!visited[ny][nx]) {
        visited[ny][nx] = true;
        q.push([ny, nx, cnt + 1]);
      }
    }
  }
  return true;
}

const places = [
  ["POOOP", "OXXOX", "OPXPX", "OOXOX", "POXXP"],
  ["POOPX", "OXPXP", "PXXXO", "OXXXO", "OOOPP"],
  ["PXOPX", "OXOXP", "OXPOX", "OXXOP", "PXPOX"],
  ["OOOXX", "XOOOX", "OOOXX", "OXOOX", "OOOOO"],
  ["PXPXP", "XPXPX", "PXPXP", "XPXPX", "PXPXP"],
];

console.log(solution(places));

/**
 * ["XOOXX", "XPOOP", "OXXPX", "XXXXX", "XXXXX"]
 */
